Source/Core/Core/AchievementManager.cpp

```diff
@@ -146,80 +154,210 @@ void AchievementManager::LoadGameByFilenameAsync(const std::string& iso_path,
           },
       .close = [](void* file_handle) { delete reinterpret_cast<FilereaderState*>(file_handle); }};
   rc_hash_init_custom_filereader(&volume_reader);
-  if (!rc_hash_generate_from_file(m_game_hash.data(), RC_CONSOLE_GAMECUBE, iso_path.c_str()))
+  m_queue.EmplaceItem([this, callback, file_path] {
+    Hash new_hash;
+    if (!rc_hash_generate_from_file(new_hash.data(), RC_CONSOLE_GAMECUBE, file_path.c_str()))
+    {
+      ERROR_LOG_FMT(ACHIEVEMENTS, "Unable to generate achievement hash from game file {}.",
+                    file_path);
+      callback(AchievementManager::ResponseType::MALFORMED_OBJECT);
+    }
+    {
+      std::lock_guard lg{m_lock};
+      m_game_hash = std::move(new_hash);
+    }
+    LoadGameSync(callback);
+  });
+}
+
+void AchievementManager::HashGame(const DiscIO::Volume* volume, const ResponseCallback& callback)
+{
+  if (!m_is_runtime_initialized)
+  {
+    ERROR_LOG_FMT(ACHIEVEMENTS,
+                  "Attempted to load game achievements without Achievement Manager initialized.");
+    callback(AchievementManager::ResponseType::MANAGER_NOT_INITIALIZED);
+    return;
+  }
+  if (m_compromised)
   {
-    ERROR_LOG_FMT(ACHIEVEMENTS, "Unable to generate achievement hash from game file.");
+    INFO_LOG_FMT(ACHIEVEMENTS, "AchievementManager is compromised and locked and cannot load a new "
+                               "game until it's restarted.");
+    OSD::AddMessage("AchievementManager is compromised and locked and cannot load a new "
+                    "game until it's restarted.",
+                    OSD::Duration::VERY_LONG, OSD::Color::RED);
     return;
   }
+  m_loading_volume = DiscIO::CreateVolume(volume->GetBlobReader().CopyReader()).get();
+  m_system = &Core::System::GetInstance();
+  struct FilereaderState
+  {
+    int64_t position = 0;
+    std::unique_ptr<DiscIO::Volume> volume;
+  };
+  rc_hash_filereader volume_reader{
+      .open = [](const char* path_utf8) -> void* {
+        auto state = std::make_unique<FilereaderState>();
+        state->volume = std::move(AchievementManager::GetInstance()->GetLoadingVolume());
```

Comment:
This `std::move` is misleading. It looks like it's transferring ownership, but actually it's doing nothing at all.